File: 654-maximum-binary-tree/654-maximum-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def getMaxIndex(nums: List[int])-> int:
    max = 0
    maxi = 0
    for i in range(0, len(nums)):
        if nums[i]> max:
            max = nums[i]
            maxi = i
    return maxi
        
def maxTree(nums: List[int]) -> Optional[TreeNode]:
    l = len(nums);
    if l == 0:
        return None
    
    if l == 1:
        return TreeNode(nums[0])
    
    maxIndex = getMaxIndex(nums)
    max = nums[maxIndex]
    
    node = TreeNode(max);
    
    left = nums[0:maxIndex]
    right = nums[maxIndex+1:l]
    
    node.left = maxTree(left)
    node.right = maxTree(right)
    
    return node
    
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        return maxTree(nums)

File: 654-maximum-binary-tree/654-maximum-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;


impl Solution {
    pub fn construct_maximum_binary_tree(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if nums.len() == 0{
            return None;
        }
        
        let max_index = nums
        .iter()
        .enumerate()
        .max_by(|(_, a), (_, b)|{ a.cmp(&b) })
        .map(|(index, _)| index)
        .unwrap();
        
        let mut node = TreeNode::new(nums[max_index]);
        
        let left_sub = nums[..max_index].to_vec();
        let right_sub = nums[max_index + 1..].to_vec();
        
        node.left = Solution::construct_maximum_binary_tree(left_sub);
        node.right = Solution::construct_maximum_binary_tree(right_sub);
        
        Some(Rc::new(
            RefCell::new(
                node
            )
        ))
        
    }
}